interpreter/prune.go

```diff
@@ -162,79 +161,124 @@ func (p *astPruner) prune(node *exprpb.Expr) (*exprpb.Expr, bool) {
+				},
+			}, true
 		}
 	case *exprpb.Expr_ListExpr:
-		newList := *node.GetListExpr()
+		elems := node.GetListExpr().GetElements()
+		newElems := make([]*exprpb.Expr, len(elems))
 		var prunedList bool
-		var prunedElem bool
-		for i, elem := range node.GetListExpr().Elements {
-			if newList.Elements[i], prunedElem = p.prune(elem); prunedElem {
+		for i, elem := range elems {
+			if newElem, prunedElem := p.prune(elem); prunedElem {
+				newElems[i] = newElem
 				prunedList = true
 			}
 		}
 		if prunedList {
-			newExpr := *node
-			newExpr.GetExprKind().(*exprpb.Expr_ListExpr).ListExpr = &newList
-			return &newExpr, true
+			return &exprpb.Expr{
+				Id: node.GetId(),
+				ExprKind: &exprpb.Expr_ListExpr{
+					ListExpr: &exprpb.Expr_CreateList{
+						Elements: newElems,
+					},
+				},
+			}, true
 		}
 	case *exprpb.Expr_StructExpr:
-		newStruct := *node.GetStructExpr()
 		var prunedStruct bool
-		var prunedEntry bool
-		for i, entry := range node.GetStructExpr().Entries {
-			newEntry := *entry
-			if newKey, pruned := p.prune(entry.GetMapKey()); pruned {
-				prunedEntry = true
-				newEntry.GetKeyKind().(*exprpb.Expr_CreateStruct_Entry_MapKey).MapKey = newKey
+		entries := node.GetStructExpr().GetEntries()
+		messageType := node.GetStructExpr().GetMessageName()
+		newEntries := make([]*exprpb.Expr_CreateStruct_Entry, len(entries))
+		for i, entry := range entries {
+			newKey, prunedKey := p.prune(entry.GetMapKey())
+			newValue, prunedValue := p.prune(entry.GetValue())
+			if !prunedKey && !prunedValue {
+				newEntries[i] = entry
 			}
-			if newValue, pruned := p.prune(entry.Value); pruned {
-				prunedEntry = true
-				newEntry.Value = newValue
+			prunedStruct = true
+			newEntry := &exprpb.Expr_CreateStruct_Entry{
+				Value: newValue,
 			}
-			if prunedEntry {
-				prunedStruct = true
-				newStruct.Entries[i] = &newEntry
+			if messageType != "" {
+				newEntry.KeyKind = &exprpb.Expr_CreateStruct_Entry_FieldKey{
+					FieldKey: entry.GetFieldKey(),
+				}
+			} else {
+				newEntry.KeyKind = &exprpb.Expr_CreateStruct_Entry_MapKey{
+					MapKey: newKey,
+				}
 			}
+			newEntries[i] = newEntry
 		}
 		if prunedStruct {
-			newExpr := *node
-			newExpr.GetExprKind().(*exprpb.Expr_StructExpr).StructExpr = &newStruct
-			return &newExpr, true
+			return &exprpb.Expr{
+				Id: node.GetId(),
+				ExprKind: &exprpb.Expr_StructExpr{
+					StructExpr: &exprpb.Expr_CreateStruct{
+						MessageName: messageType,
+						Entries:     newEntries,
+					},
+				},
+			}, true
 		}
 	case *exprpb.Expr_ComprehensionExpr:
-		if newIterRange, pruned := p.prune(node.GetComprehensionExpr().IterRange); pruned {
-			newExpr := *node
-			newCompre := *newExpr.GetComprehensionExpr()
-			newCompre.IterRange = newIterRange
-			newExpr.GetExprKind().(*exprpb.Expr_ComprehensionExpr).ComprehensionExpr = &newCompre
-			return &newExpr, true
+		compre := node.GetComprehensionExpr()
+		if newRange, pruned := p.prune(compre.GetIterRange()); pruned {
```

Comment:
Add a comment explaining why it's sufficient to only test the `iter_range` but not any of the other sub-expressions.